beefy-primitives/src/commitment.rs

```diff
@@ -95,6 +99,136 @@ impl<TBlockNumber, TPayload, TSignature> SignedCommitment<TBlockNumber, TPayload
+	fn pack(signed_commitment: &'a SignedCommitment<TBlockNumber, TPayload, TSignature>) -> Self {
+		let SignedCommitment { commitment, signatures } = signed_commitment;
+		let signatures_len = signatures.len() as u32;
+		let mut signatures_from: BitField = vec![];
+		let mut signatures_compact: Vec<&TSignature> = vec![];
+
+		for signature in signatures {
+			match signature {
+				Some(value) => signatures_compact.push(value),
+				None => (),
+			}
+		}
+
+		let bits: Vec<u8> = signatures.iter().map(|x| if x.is_some() { 1 } else { 0 }).collect();
+		let chunks = bits.chunks(CONTAINER_BIT_SIZE);
+		for chunk in chunks {
+			let mut iter = chunk.iter().copied();
+			let mut v = iter.next().unwrap() as u8;
+
+			for bit in iter {
+				v <<= 1;
+				v |= bit as u8;
+			}
+
+			signatures_from.push(v);
+		}
+
+		Self {
+			commitment,
+			signatures_from,
+			signatures_len,
+			signatures_compact,
+		}
+	}
+}
+
+impl<TBlockNumber, TPayload, TSignature> Encode for SignedCommitment<TBlockNumber, TPayload, TSignature>
+where
+	TSignature: Encode,
+	TBlockNumber: Encode,
+	TPayload: Encode,
+{
+	fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+		let temp = CompactSignedCommitment::pack(self);
+		temp.using_encoded(f)
+	}
+}
+
+impl<TBlockNumber, TPayload, TSignature> CompactSignedCommitment<Commitment<TBlockNumber, TPayload>, TSignature>
+where
+	TBlockNumber: Decode,
+	TPayload: Decode,
+	TSignature: Decode,
+{
+	/// Unpacks a `CompactSignedCommitment` into the uncompressed `SignedCommitment` form.
+	fn unpack(
+		temporary_signatures: CompactSignedCommitment<Commitment<TBlockNumber, TPayload>, TSignature>,
+	) -> SignedCommitment<TBlockNumber, TPayload, TSignature> {
+		let CompactSignedCommitment {
+			commitment,
+			signatures_from,
+			mut signatures_len,
+			signatures_compact,
+		} = temporary_signatures;
+		let mut bits: Vec<u8> = vec![];
+		let last_block = signatures_len % CONTAINER_BIT_SIZE as u32;
+		for block in signatures_from {
+			let start_bit = if signatures_len > last_block {
+				0
+			} else {
+				CONTAINER_BIT_SIZE - last_block as usize
+			};
+
+			for bit in start_bit..CONTAINER_BIT_SIZE {
+				let bit_position = CONTAINER_BIT_SIZE - bit - 1;
+				bits.push(block >> bit_position & 1);
+				signatures_len -= 1;
+			}
+		}
+
+		let mut next_signature = signatures_compact.into_iter();
+		let signatures: Vec<Option<TSignature>> = bits
+			.iter()
+			.map(|&x| if x == 1 { next_signature.next() } else { None })
+			.collect();
+
+		SignedCommitment { commitment, signatures }
+	}
+}
+
+impl<TBlockNumber, TPayload, TSignature> Decode for SignedCommitment<TBlockNumber, TPayload, TSignature>
+where
+	TBlockNumber: Decode,
+	TPayload: Decode,
+	TSignature: Decode,
+{
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		let temp = CompactSignedCommitment::decode(input)?;
+		Ok(CompactSignedCommitment::unpack(temp))
```

Comment:
I just realised that we already have chains that store signed commitments in the database (Rococo & Wococo).
We may either restart them or we could introduce a backward compatible reading here temporarily (until they are restarted).

@andresilva @adoerr wdyt? 

A backward compatible suggestion would look like this (note we should add a test for this):

```suggestion
		let temp = match CompactSignedCommitment::decode(input) {
		  Ok(temp) => Ok(CompactSignedCommitment::unpack(temp)),
		  Err(_) => {
		     type OldSignedCommitment = (Commitment, Vec<Option<Signature>>);
		      let (commitment, signatures) = OldSignedCommitment::decode(input)?;
		      log::warn!("Decoded `SignedCommitment` in old format, this is not supported and will be removed soon. Please restart or migrate your chain.");
		      Ok(SignedCommitment { commitment, signatures })
		  }
		};
```
I guess we might need to do some magic to clone the `input` though, but it should be doable.

Reply:
After discussing with @andresilva and @adoerr we've figured that we can work around backward compatibility issue. So please disregard the comment.